include/drivers/dai.h

```diff
@@ -0,0 +1,717 @@
+	 * DAI_TRIGGER_STOP. This trigger can be used in RUNNING state only and
+	 * at first changes the interface state to STOPPING. When all TX blocks
+	 * are transmitted the state is changed to READY.
+	 */
+	DAI_TRIGGER_DRAIN,
+	/** @brief Discard the transmit / receive queue.
+	 *
+	 * Stop the transmission / reception immediately and discard the
+	 * contents of the respective queue. This trigger can be used in any
+	 * state other than NOT_READY and changes the interface state to READY.
+	 */
+	DAI_TRIGGER_DROP,
+	/** @brief Prepare the queues after underrun/overrun error has occurred.
+	 *
+	 * This trigger can be used in ERROR state only and changes the
+	 * interface state to READY.
+	 */
+	DAI_TRIGGER_PREPARE,
+	/** @brief Reset
+	 *
+	 * This trigger frees resources.
+	 */
+	DAI_TRIGGER_RESET,
+};
+
+/** DAi properties used for binding with dma and other things */
+enum dai_property {
+	/* fifo id */
+	DAI_PROPERTY_FIFO = 0,
+	/* handshake id */
+	DAI_PROPERTY_HANDSHAKE,
+	DAI_PROPERTY_INIT_DELAY,
+};
+
+/** @struct dai_legacy_params
+ * @brief Interface configuration options.
+ *
+ * Copied exactly from existing i2s interface to provide integration path for 
+ * legacy zephyr drivers.
+ *
+ * Memory slab pointed to by the mem_slab field has to be defined and
+ * initialized by the user. For DAI driver to function correctly number of
+ * memory blocks in a slab has to be at least 2 per queue. Size of the memory
+ * block should be multiple of frame_size where frame_size = (channels *
+ * word_size_bytes). As an example 16 bit word will occupy 2 bytes, 24 or 32
+ * bit word will occupy 4 bytes.
+ *
+ * Please check Zephyr Kernel Primer for more information on memory slabs.
+ *
+ * @remark When DAI data format is selected parameter channels is ignored,
+ * number of words in a frame is always 2.
+ *
+ * @param word_size Number of bits representing one data word.
+ * @param channels Number of words per frame.
+ * @param format Data stream format as defined by DAI_FMT_* constants.
+ * @param options Configuration options as defined by DAI_OPT_* constants.
+ * @param frame_clk_freq Frame clock (WS) frequency, this is sampling rate.
+ * @param mem_slab memory slab to store RX/TX data.
+ * @param block_size Size of one RX/TX memory block (buffer) in bytes.
+ * @param timeout Read/Write timeout. Number of milliseconds to wait in case TX
+ *        queue is full or RX queue is empty, or 0, or SYS_FOREVER_MS.
+ */
+struct dai_legacy_params {
+	uint8_t word_size;
+	uint8_t channels;
+	dai_fmt_t format;
+	dai_opt_t options;
+	uint32_t frame_clk_freq;
+	struct k_mem_slab *mem_slab;
+	size_t block_size;
+	int32_t timeout;
+};
+
+/** Main dai config struct */
+struct dai_config {
+	enum dai_type type;
+	uint32_t dai_index;
+
+	/* generic dai config members */
+	dai_fmt_t format;
+	uint32_t channels;
+	uint32_t frame_clk_freq;
+	uint8_t group_id;
+	uint8_t flags;
+
+	/* optional dai specific params */
+	const void *params;
```

Comment:
So the thing is that some of the config structs are huge (for example dmic). Whatever way I define those (typed) it is kind of  ugly, for example the dmic structure has 3 levels of nested structs in arrays. I needed actually to increase some heap size to fit those in. So I would keep some kind of private data pointer even if it might cause some issues.

Reply:
We can do a 
```c
union {
    struct intel_dmic *intel_dmic;
    struct intel_ssp *intel_ssp;
    struct nxp_i2s *nxp_i2s;
    /* and so on ... */
};
```
The key point is the data is bespoke configuration that cannot easily be defined by a generic or common API, but is abstracted here with a void * or a union. 

This data comes from topology or ACPI on Intel platforms, but could also come from DT on non Intel platforms (or in non host mode).

Reply:
@juimonen How this config relates to the bespoke option we pass later. It seems we are passing two opaque pointers in the same API, is this intentional?